### PR TITLE
Fantomas: detect and run specified version

### PR DESCRIPTION
**Pull request by DedSec256**
image

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FantomasHost.fs [hunk 1]
```diff
@@ -20,20 +20,29 @@ module internal Reflection =
...
 
     let toEditorConfigName name = $"{fSharpEditorConfigPrefix}{StringUtil.MakeUnderscoreCaseName(name)}"
 
     let isConnectionAlive () =
         isNotNull connection && connection.IsActive
 
+    let terminateConnection () =
+        if isConnectionAlive () then formatterHostLifetime.Terminate()
+
+    // Formatting is called on UI thread, so we don't need to synchronize it
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FantomasHost.fs:37]: I think it worth it to either synchronize it or to assert we only access Fantomas on the main thread. Asserting is definitely a better way to make it safe than a comment. Otherwise it'd be a hard to spot issue if something changes in future (and I do hope we'll have less things working on the main thread like this).

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FSharpCodeStylePage.fs [hunk 2]
```diff
@@ -20,7 +20,7 @@ type FSharpCodePreviewPreparator() =
 
 [<FormattingSettingsPresentationComponent>]
 type FSharpCodeStylePageSchema(lifetime, smartContext, itemViewModelFactory, container, settingsToHide) =
-    inherit CodeStylePageSchema<FSharpFormatSettingsKey, FSharpCodePreviewPreparator>(lifetime, smartContext,
+    inherit IndentStylePageSchema<FSharpFormatSettingsKey, FSharpCodePreviewPreparator>(lifetime, smartContext,
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FSharpCodeStylePage.fs:23]: Could you tell why this change is needed, please? (Just curious.)

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    | UnsupportedVersion
+    | SelectedButNotFound
+
+type FantomasVersion =
+    | Bundled
+    | SolutionDotnetTool
+    | GlobalDotnetTool
+
+type FantomasRunSettings =
+    { Version: FantomasVersion * string
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:39]: The purpose of the `string` item is very hard to guess. Could you redesign this part, so it's easier to understand, please? Could `Version` type better fit the version role?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+type FantomasDetector(lifetime, settingsProvider: FSharpFantomasSettingsProvider,
+                      dotnetToolsTracker: NuGetDotnetToolsTracker) =
+    let [<Literal>] fantomasToolPackageId = "fantomas-tool"
+    let [<Literal>] fantomasPackageId = "fantomas"
+    let minimalSupportedVersion = Version(MinimalSupportedVersion)
+    let notificationSignal = Signal<FantomasDiagnosticNotification>()
+    let rwLock = JetFastSemiReenterableRWLock()
+
+    let calculateGlobalFantomasToolPath version =
+        Path.Combine(CliFolderPathCalculator.ToolPackageFolderPath, fantomasToolPackageId, version, fantomasToolPackageId, version, "tools", "netcoreapp3.1", "any", "Fantomas.dll")
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:58]: Could you split this into getting the directory and getting the tool inside it? Just to make it a bit easier to follow.

Also, any reason not to use `FileSystemPath`-derived types?

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:58]: Is `netcoreapp3.1` the only target framework Fantomas uses?

@nojaf Do you know if it can change in foreseeable future?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+                    notificationsFired.Add(version) |> ignore
+                    { Event = status; Version = version; FallbackVersion = versionToRun } :: warnings
+
+                versionToRun, warnings
+            | false, _ -> calculateVersionRec nextVersionToSearch warnings
+
+        calculateVersionRec version []
+
+    let rec calculateRunSettings setting =
+        let startToSearchVersion =
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:96]:
```suggestion
        let searchStartVersion =
```

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        | FantomasVersionSettings.GlobalDotnetTool ->
+            if versionsData.ContainsKey(GlobalDotnetTool) then () else
+            versionsData.Add(GlobalDotnetTool, { Version = GlobalDotnetTool, ""; Path = null; Status = SelectedButNotFound })
+
+        | _ -> ()
+
+    do
+        if isNull settingsProvider || isNull dotnetToolsTracker then () else
+
+        settingsProvider.Version.Change.Advise(lifetime, fun x ->
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:164]: What do you think about using `fantomasSettingsProvider` name? Or `fantomasVersionProvider`? It'd one less thing to guess when looking at this constructor.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        fireNotifications()
+        let { Path = path } as versionToRun = versionToRun.Value
+        try runAction path
+        with _ ->
+            versionToRun.Status <- FailedToRun
+            recalculateState settingsProvider.Version.Value
+            x.TryRun(runAction)
+
+    member x.GetSettings() =
+        let settings = Dictionary<_, _>()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:190]: Any reason to specify the wildcard types? It doesn't seem required here. I think it's been asked in previous PRs.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+            .AddColumnElementsToNewRow(BeSizingType.Fit, true, [|"Fantomas version".GetBeLabel() :> BeControl; beComboBoxFromEnum.WithMinSize(BeControlSizeFixed(BeControlSizeType.FIT_TO_CONTENT, BeControlSizeType.FIT_TO_CONTENT), lifetime)|])
+            .AddColumnElementsToNewRow(BeSizingType.Fit, false, [|"".GetBeLabel() :> BeControl; validationLabel |])
+
+    do
+        use indent = this.Indent()
+
+        this.AddControl((fun (key: FSharpFantomasOptions) -> key.Version), createComboBox) |> ignore//.DisableDescendantsWhenDisabled(this.Lifetime)
+
+        this.AddCommentText("To use a specified Fantomas version, install it globally via 'dotnet tool install fantomas-tool'\n" +
+                            "or specify it in dotnet-tools.json file in the solution directory. Supported Fantomas versions: 3.2 and later.")
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:287]: Any reason the version is inlined in some places and uses string interpolation in others?

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:287]: Btw, do we need to show the version here at all? Is it common enough to install such old versions via `dotnet`, so it's worth it to show it to every user opening this page?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs (hunk 4 above)
Comment (nojaf) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:58]: Yes, in 4.7 we will most likely target `net6.0`.

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:58]: Thanks! That means we should not hard-code it @DedSec256.

File: ReSharper.FSharp/src/FSharp.Fantomas.Host/FSharp.Fantomas.Host.csproj [hunk 9]
```diff
@@ -22,6 +22,7 @@
     </ItemGroup>
 
     <ItemGroup>
+      <!-- update BundledVersion in FantomasOptionsPage.fs/FantomasVersions if changed -->
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Fantomas.Host/FSharp.Fantomas.Host.csproj:25]: We can do this if we run Fantomas

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs (hunk 4 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:58]: This code is a temporary workaround and will be removed when we add getting global tool paths to the platform since this logic is more non-trivial and different from how we get paths for local tools.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs (hunk 8 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:287]: > Any reason the version is inlined in some places and uses string interpolation in others?

forgot about this place, thanks

> Btw, do we need to show the version here at all? Is it common enough to install such old versions via dotnet, so it's worth it to show it to every user opening this page?

yes, I think we can not follow the prototype here and really not show.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FantomasHost.fs (hunk 1 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCleanup/FantomasHost.fs:37]: Generally speaking, we have all the necessary synchronization in `FantomasDetector`, through which we notify about a Fantomas version change and run Fantomas. If we wrap the entire body of the `connect` function in fantomasDetector.Run(...), then have all the necessary synchronization regardless of the thread calling formating.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs (hunk 3 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs:39]: We often want to have a string representation of the version and use `Version` just to validate it. Perhaps a `Version option` would be better since the found version can be invalid.

File: ReSharper.FSharp/src/FSharp.Fantomas.Host/FSharp.Fantomas.Host.csproj (hunk 9 above)
[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Fantomas.Host/FSharp.Fantomas.Host.csproj:25]: looks like we can use something like this https://www.eidias.com/blog/2021/1/12/custom-attributes-in-net-core-csproj-file

File: ReSharper.FSharp/src/FSharp.Fantomas.Protocol/src/FantomasProcessFactory.cs [hunk 10]
```diff
@@ -24,13 +24,13 @@ public class FantomasProcessFactory
...
 
-    public FantomasProcess Create(Lifetime lifetime)
+    public FantomasProcess Create(Lifetime lifetime, VirtualFileSystemPath path = null)
     {
       return new FantomasProcess(lifetime,
         myLogger,
         myShellLocks,
         mySolutionProcessStartInfoPatcher,
-        JetProcessRuntimeRequest.CreateInternalRuntime());
+        JetProcessRuntimeRequest.CreateCore(), path);
```
Comment (DedSec256) [inline ReSharper.FSharp/src/FSharp.Fantomas.Protocol/src/FantomasProcessFactory.cs:33]: Since we are running fantomas always on .NET Core anyway
